File: src/main/java/com/murex/tbw/report/ReportGenerator.java
package com.murex.tbw.report;

import com.murex.tbw.MainRepository;
import com.murex.tbw.purchase.Invoice;
import com.murex.tbw.purchase.PurchasedBook;
import com.murex.tbw.storage.Repository;

import java.util.List;
import java.util.Map;

public class ReportGenerator {

    private final Repository repository = MainRepository.configuredRepository();

    public double getTotalAmount() {
        Map<Integer, Invoice> invoiceMap = repository.getInvoiceMap();
        return invoiceMap.values().stream().mapToDouble(Invoice::computeTotalAmount).sum();
    }

    public int getTotalSoldBooks() {
        Map<Integer, Invoice> invoiceMap = repository.getInvoiceMap();
        return invoiceMap.values().stream().mapToInt(invoice -> {
            List<PurchasedBook> purchasedBooks = invoice.getPurchasedBooks();
            return purchasedBooks.stream().mapToInt(PurchasedBook::getQuantity).sum();
        }).sum();
    }

    public long getNumberOfIssuedInvoices() {
        Map<Integer, Invoice> invoiceMap = repository.getInvoiceMap();
        return invoiceMap.values().size();
    }
}

File: typescript/src/com/murex/tbw/report/ReportGenerator.ts
import { Repository } from "../storage/Repository";
import { MainRepository } from "../MainRepository";
import { toUSD } from "../finance/CurrencyConverter";

export class ReportGenerator {
  private readonly repository: Repository = MainRepository.configuredRepository();

  public getTotalAmount(): number {
    const invoiceMap = this.repository.getInvoiceMap();
    let totalAmount = 0.0;

    invoiceMap.valueSeq().forEach((invoice) => {
      // BUG: There was a bug with the below line of code
      // totalAmount += invoice.computeTotalAmount();

      // FIX: The above bug was fixed by the below 2 lines of code
      totalAmount += toUSD(
        invoice.computeTotalAmount(),
        invoice.country.currency
      );
    });
    return this.getRoundedAmount(totalAmount);
  }

  private getRoundedAmount(totalAmount: number): number {
    return Math.round(totalAmount * 100.0) / 100.0;
  }

  getTotalSoldBooks() {
    const invoiceMap = this.repository.getInvoiceMap();
    let totalSoldBooks = 0;

    invoiceMap.valueSeq().forEach((invoice) => {
      invoice.purchasedBooks.forEach((purchasedBook) => {
        totalSoldBooks = totalSoldBooks + purchasedBook.quantity;
      });
    });
    return totalSoldBooks;
  }

  getNumberOfIssuedInvoices() {
    const invoiceMap = this.repository.getInvoiceMap();
    return invoiceMap.size;
  }
}
